### ISSUE TITLE
PKI secret engine lacks necessary operation to sign a new CSR for the intermediate issuing certificate

### ISSUE THREAD

**Issue by maxb** [hashicorp/vault#10815]
**Describe the bug**
The PKI secret engine provides no way to retrieve a new signed CSR for an existing intermediate CA certificate. By not providing this option, a trap is laid for the less experienced Vault operator, which they will only discover as their intermediate CA certificate approaches expiry, and they would like to renew it.

You could consider this to be a feature request for such an operation to be added. However, I suggest it merits consideration as a bug, because of how it can entrap a less experienced operator.

This bug is a specific instance of the general discussion on this topic in #2581 

**To Reproduce**
1. Configure a new PKI secrets engine instance using the `/intermediate/generate/internal` endpoint workflow described at https://www.vaultproject.io/docs/secrets/pki
2. Put this PKI secrets engine into production, and use it for a while until it is nearing expiry
3. Realise that you have backed yourself into a corner, and cannot extend the lifetime of this CA in a compatible manner

**Expected behavior**
The PKI secrets engine should have an API operation, which might be named `/intermediate/generate/csr-only`, which would behave in an identical manner to `/intermediate/generate/internal`, except that it would NOT reset any CA configuration, and would just proceed directly to constructing, signing, and returning a CSR using the CA private key already stored within the secrets engine.

With this operation available, there would then be a readily explainable workflow for renewing a PKI secrets engine intermediate (in the case where the same private key continues to be used) - effectively this is solving the easy part of #2581, decoupled from the far harder issue of what the desired behaviour should be if you have to rotate the private key too.

**Comment by HridoyRoy** [comment]
Hi @maxb , thank you for filing this issue. I'll raise it to the team and report back.

**Comment by jefferai** [comment]
Hi @maxb ,

The reason this isn't (currently) offered is that generally speaking intermediates are "disposable" in the sense that when they expire, you don't renew the certificate; rather, you create a new intermediate with a new key. This allows scoping the lifetime of the intermediate to the organization policy around the lifetime of the key. Intermediates are more heavily used than roots (which are usually only exposed to sign intermediate CSRs, and are otherwise kept securely locked away), so it is best practice to rotate their key regularly.

Because the root is the same, and the intermediate is only used to construct a verification path to the root, creating new keys and having those signed by the CA does not disrupt users at all.

Since in most cases organizations will scope the lifetime of the intermediate to the allowed key lifetime, this is not a request that generally comes up. I can't speak for the Vault product team in terms of whether they will implement this request, I just wanted to provide some context as to why it doesn't exist right now.

**Comment by maxb** [comment]
Hi @jefferai ,

I do agree that what you say is what *should* be the case in an ideal scenario.

However, Vault makes it relatively easy to design yourself into a corner where renewing the intermediate is a useful operation:

1. Create a new PKI secrets engine. Accept the default mount path of "pki/".
2. Get its intermediate certificate signed by a corporate root CA.
3. Build a large ecosystem of applications, maintained by diverse internal teams, which obtain certificates from the pki/ secrets engine.

Then, nearly 2 years later, discover your initial 2-year intermediate certificate is approaching expiry. At this point, your options are rather limited...

If you choose to set up a new intermediate at a different mount path in Vault, each application needs to be repointed at new URLs. This is somewhere between inelegant and infeasible depending on the number of applications.

If you choose to re-key the existing PKI secrets engine then there are multiple downsides:
* You break the previous CA's CRL and AIA URLs
* You take certificate issuance offline whilst the new intermediate CSR is sent to the root CA and signed

If you can get your hands on a CSR for the existing key, though, you can extend the validity of the intermediate CA, and buy time to consider what your next steps should be.

I do have an idea for a more complex solution that would allow re-keying, which I'm going to write up in a new issue.

However, the feature described here should have a very low cost to implement - it's basically just an if-statement to skip part of what `/intermediate/generate/internal` already does, and has the potential to make life easier for people who have fallen into this PKI design trap.

**Comment by jefferai** [comment]
> You break the previous CA's CRL and AIA URLs

But why not distribute those from a different endpoint? Those are plain HTTP calls that can be handled by any server.

This is best practice in large environments anyways, to keep this traffic on a simple HTTP server and away from a complex CA application.

You keep those URLs valid, set up a new intermediate pointing to a new external HTTP location, and everything works.

**Comment by maxb** [comment]
> > You break the previous CA's CRL and AIA URLs
> 
> But why not distribute those from a different endpoint? Those are plain HTTP calls that can be handled by any server.
> 
> This is best practice in large environments anyways, to keep this traffic on a simple HTTP server and away from a complex CA application.
>
> You keep those URLs valid, set up a new intermediate pointing to a new external HTTP location, and everything works.

I agree with you ... but what if you're not in a large environment, or you didn't think you were in a large enough environment to bother when you started issuing certificates, and need to continue to support them, or you just naively followed https://www.vaultproject.io/docs/secrets/pki which gives an example of pointing these directly at Vault?

Remember that Vault makes a PKI secrets engine's `/ca` and `/crl` available unauthenticated and in bare formats conducive to being used as CRL and AIA URLs directly, so it is sending a strong message that this is an OK thing to do until you need to scale out.

And furthermore - even if you do move your CRL and AIA off Vault, this still doesn't address the issue that re-keying a PKI secrets engine in place, will cause it to experience downtime whilst the CSR is transported to the root CA, signed, and the resulting certificate installed back into Vault.



I'm not claiming that the feature proposed here is part of a start-to-finish best practice solution. I am claiming it's an expedient, low engineering cost way to make Vault PKI a more friendly and accomodating system for Vault PKI administrators who know more now, than they did when they initially set up their production PKI.

**Comment by cipherboy** [comment]
@maxb This should now be addressed in the #15277 and doc'd in #15238 -- you'll see a new `intermediate/cross-sign` endpoint and also the `issuers/generate/intermediate/existing` endpoint (equivalent to `intermediate/generate/existing`), which allows for creation of CSRs with existing key material in Vault. The motivation for this feature set was originally rotation of roots and intermediates, so I do believe it should address the issue.

Feel free to comment or reopen if you've got questions or issues.
